src/internet_identity/src/anchor_management.rs

```diff
@@ -109,6 +109,36 @@ pub fn update(user_number: AnchorNumber, device_key: DeviceKey, device_data: Dev
     );
 }
 
+pub fn replace(anchor_number: AnchorNumber, device_key: DeviceKey, device_data: DeviceData) {
+    let mut anchor = state::anchor(anchor_number);
+    // must be called before the first await because it requires caller()
+    trap_if_not_authenticated(&anchor);
+
+    anchor.remove_device(&device_key).unwrap_or_else(|err| {
+        trap(&format!(
+            "failed to replace device of anchor {}: {}",
+            anchor_number, err
+        ))
+    });
+    let new_device = Device::from(device_data);
```

Comment:
I've asked this a thousand times already and apologize for that, but can you confirm that a canister call is pretty much a transaction? i.e. if this line traps then all changes to orthogonal persistence and stable memory will be rolled back and the `remove_device` call above will be rollbacked too?

Reply:
Yes, it is. All heap changes and stable memory changes will be rolled back.